### ISSUE TITLE
Add getScreenBBox() method to SVGGraphicsElement interface

### ISSUE THREAD

**Issue by jarek-foksa** [w3c/svgwg#76]
There is no way to determine the bounding box of an SVG element in the client space.

The generic `element.getBoundingClientRect` method, unlike `element.getBBox`, does not allow you to specify whether to include stroke and markers in the result. It also does not return a geometric bbox, but rather a user bbox that has been transformed to the client space, which is not what you would want in most scenarios.

Therefore, I would propose to add `getScreenBBox` method to the `SVGGraphicsElement` interface. The method would take same arguments as existing `getBBox` method.

I know it's technically "client" rather than "screen", but I would stay with the term "screen" for the sake of consistency with `getCTM`/`getScreenCTM`.

**Comment by BigBadaboom** [comment]
Or alternatively, perhaps, add a new property to `SVGBoundingBoxOptions`?

```
dictionary SVGBoundingBoxOptions {
  boolean fill = true;
  boolean stroke = false;
  boolean markers = false;
  boolean clipped = false;
  boolean applyTransforms = false;
};
```

If `applyTransforms` is true, then the bounding box that is calculated takes into account all transforms on the current and ancestor elements.

**Comment by jarek-foksa** [comment]
@foolip: In order to determine the geometric bbox of a path in the client space you can't just transform paths' user (local) bbox to the client space, this might produce wrong results when the path or its ancestors are transformed.

Instead, you have to determine the coordinates of each path node point, then transform each point to the client space and finally use that to determine the bbox.

**Comment by fsoder** [comment]
Ah, yes, I missed that part of the issue. Computing the "precise" bounding box in "screen-space" will be costly even if "querying the render tree" though - unless this happen to be stored already (it isn't in Blink for example. There's a cost associated with maintaining this kind of "absolute" state - if there weren't, gBCR et al would likely already return these tight fitting bboxes.)

**Comment by AmeliaBR** [comment]
Yes, we already use the simplified approach in other cases when combining transforms and BBox calculations, precisely for the issue of computational complexity.

Currently, if you compute the BBox of a rotated path, or of a `<g>` that contains a rotated path, you get the bounds in the parent's coordinate space of the path's rotated BBox, not tight bounds for the rotated path.

[This CodePen by Sarah Drasner demonstrates the effect clearly](http://codepen.io/sdras/pen/MwxRBL?editors=0010). 

Therefore, for compatibility, if a `getScreenBBox()` method was introduced, it would probably follow the steps in [@fsoder's comment](https://github.com/w3c/svgwg/issues/76#issuecomment-291010647), although I'd recommend letting the method take the options object as a parameter, rather than assuming certain values.

**Comment by jarek-foksa** [comment]
@AmeliaBR 

From my testing `pathElement.getBBox()` returns a proper [geometric minimum bounding box](https://en.wikipedia.org/wiki/Minimum_bounding_box) in the user space. I.e. the smallest possible rectangle inside which the path fits in, computed using raw path data (without applying any transforms).

That gist is demonstrating the problem I was talking about earlier - you can't just take a bounding box in the user space and transform it's coordinates to some other space - this will produce a bounding box of another bounding box which is not useful and wrong if you stick to the mathematical definition.

Making `getScreenBBox()` behave this way would in my opinion be inconsistent with how `getBBox()` works and redundant because there is already `element.getScreenBBox()` which does it.

**Comment by BigBadaboom** [comment]
There is a *very* common request on Stack Overflow to get the bounds of an element **including** its own transform.  IOW returning `{x:50, y:0, width:100, height:100}` for

    <rect width="100" height="100" transform="translate(50,0)"/>

So it would be good to consider this use case in any solution.

So, using my previous suggestion, for example:

    dictionary SVGBoundingBoxOptions {
      boolean fill = true;
      boolean stroke = false;
      boolean markers = false;
      boolean clipped = false;
      boolean selfTransform = false;
      boolean ancestorTransforms = false;
    };

**Comment by fsoder** [comment]
Having `selfTransform` and `ancestorTransforms` means you have the possibility of `{ selfTransform: false, ancestorTransforms: true }` which is a bit of an oddity. So an enumeration (`enum ApplyTransforms { "none", "self-only", "self-and-ancestors" }; ... ApplyTransforms applyTransforms = "none" ...`) or even a (nullable) Element might be a better choice.

**Comment by BigBadaboom** [comment]
Yes.  Good point.

**Comment by paradisaeidae** [comment]
Blake/ OSUblake has a great writeup of this issue here: https://greensock.com/forums/topic/13681-svg-gotchas/?page=2&tab=comments#comment-72060

His codepen of a workaround is excellent!
